### PR TITLE
Add spirit & stamina consumption

### PR DESCRIPTION
**Pull request by shuabritze**
Updating is just based on field tickCount and not an async task, might be better to have long-running regens be in another thread.

### CODE REVIEW

**Comment by mettaursp** [comment]
Ayyy here's the regen

> Updating is just based on field tickCount and not an async task, might be better to have long-running regens be in another thread.

Nah I don't think a full extra thread is necessary for something as simple & low cost as regen. Adding threads is a big ask with a big maintenance cost & generally only worth it when the cost is big enough to halt the main thread. Every extra thread is extra synchronization that needs to be done, which would make it more costly than anything you save in the case of something like this.

The implementation here is fine as is.

File: Maple2.Server.Game/Model/Skill/SkillRecord.cs [hunk 1]
```diff
@@ -39,6 +40,20 @@ public class SkillRecord {
...
+        return 0;
+    }
+
+    public int GetStaCost() {
+        if (Metadata.Data.Consume.Stat.TryGetValue(BasicAttribute.Stamina, out long staminaCost)) {
+            return (int) staminaCost;
+        }
+        return 0;
+    }
+
```
Comment (Zintixx) [inline Maple2.Server.Game/Model/Skill/SkillRecord.cs:56]: I could be wrong here but isn't getting the Sp or Stamina Cost only a one time occurance - through skill use from a player? I would think we could just move this and merge it into the SkillHandler UseSkill function

File: Maple2.Server.Game/PacketHandlers/SkillHandler.cs [hunk 2]
```diff
@@ -128,6 +128,33 @@ private enum SubCommand : byte {
...
+                Logger.Error("Not enough stamina to cast skill: {SkillId},{Level}", skillId, level);
+                return;
+            }
+        }
+
+        session.Player.ConsumeSp(spiritCost);
+        session.Player.ConsumeStamina(staminaCost);
+        session.Field.Broadcast(StatsPacket.Update(session.Player, [BasicAttribute.Spirit, BasicAttribute.Stamina]));
+
+        session.ConditionUpdate(ConditionType.skill, 1, codeLong: skillId);
```
Comment (Zintixx) [inline Maple2.Server.Game/PacketHandlers/SkillHandler.cs:156]: `session.ConditionUpdate(ConditionType.skill, codeLong: skillId);`

You'll also need to handle this in Server.Game/Util/ConditionUtil

File: Maple2.Server.Game/Model/Skill/SkillRecord.cs (hunk 1 above)
Comment (mettaursp) [reply Maple2.Server.Game/Model/Skill/SkillRecord.cs:56]: Yeah should be I think.

**Review by Zintixx — APPROVED** [review]

**Review by AngeloTadeucci — APPROVED** [review]

**Pull request merged**